extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -308,19 +308,22 @@ public void stop(boolean stopGracefully)
       // until the tasks have acknowledged or timed out. We want this behavior when we're explicitly shut down through
       // the API, but if we shut down for other reasons (e.g. we lose leadership) we want to just stop and leave the
       // tasks as they are.
-      if (stopGracefully) {
-        log.info("Stopping gracefully, signalling managed tasks to complete and publish");
-        synchronized (stopLock) {
+      synchronized (stopLock) {
+        if (stopGracefully) {
+          log.info("Posting GracefulShutdownNotice, signalling managed tasks to complete and publish");
+          notices.add(new GracefulShutdownNotice());
+        } else {
+          log.info("Posting ShutdownNotice");
           notices.add(new ShutdownNotice());
-          while (!stopped) {
-            stopLock.wait();
-          }
         }
-        log.info("Shutdown notice handled");
+
+        while (!stopped) {
+          stopLock.wait();
```

Comment:
Hm, this is only an issue with a graceful shutdown where it tries to synchronize a publish of the tasks; if we quit before the supervisor is done handling all the tasks (either getting acknowledgment that they're publishing or force killing them if they're not responding), the worst case is that we could wind up with some tasks that have been paused but haven't yet been told to publish which will hang around doing nothing until the next supervisor starts up and takes control of them, or they might not have been paused and may just keep reading until someone does something about them.